examples/nivc_dynamic_phi.rs

```diff
@@ -0,0 +1,360 @@
+use std::marker::PhantomData;
+
+use bellpepper::gadgets::{blake2s::blake2s, multipack::pack_bits, sha256::sha256};
+use bellpepper_core::{boolean::Boolean, num::AllocatedNum, ConstraintSystem, SynthesisError};
+use blake2s_simd::Params;
+use ff::{Field, PrimeField, PrimeFieldBits};
+use rand::Rng;
+use rand_core::OsRng;
+use sha2::{Digest, Sha256};
+
+use nova_snark::{
+  parameters::PublicParams,
+  supernova::{NonUniformCircuit, RecursiveSNARK},
+  traits::{
+    circuit_supernova::{StepCircuit, TrivialSecondaryCircuit},
+    Group,
+  },
+};
+
+const NUM_STEPS: usize = 10;
+const NUM_BYTES: usize = 10;
+
+#[derive(Clone, Debug)]
+struct SHACircuit<F: PrimeField> {
+  next_pc: bool,
+  _p: PhantomData<F>,
+}
+
+impl<F: PrimeField> SHACircuit<F> {
+  fn new(next_pc: bool) -> Self {
+    Self {
+      next_pc,
+      _p: PhantomData,
+    }
+  }
+}
+
+impl<F: PrimeField + PrimeFieldBits> StepCircuit<F> for SHACircuit<F> {
+  fn arity(&self) -> usize {
+    1
+  }
+
+  fn synthesize<CS: ConstraintSystem<F>>(
+    &self,
+    cs: &mut CS,
+    _pc: Option<&AllocatedNum<F>>,
+    z: &[AllocatedNum<F>],
+  ) -> Result<(Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError> {
+    let preimage = &z[0];
+
+    let preimage_bits = preimage.to_bits_le_strict(cs.namespace(|| "sha_preimage_bits"))?;
+
+    let (preimage_bits, _) = preimage_bits.split_at(8 * NUM_BYTES);
+
+    let preimage_bits = preimage_bits
+      .chunks(8)
+      .flat_map(|chunk| {
+        let mut chunk = chunk.to_vec();
+        chunk.reverse();
+        chunk
+      })
+      .collect::<Vec<_>>();
+
```

Comment:
I think you could just pack the 80 bits you want into the preimage here — same as you do for the digest below. That would save the need for the strict decomposition and to constrain the unused bits (so savings of 200+ constraints).

Reply:
This only works if the inputs will always be 80 bits, but I think that is the case for all the internal inputs, if all of your outputs are 80 bits (stuffed into a field element). So if you similarly restrict the initial input, then this would work.